### ISSUE TITLE
Enabling the external memory in R and Windows

### ISSUE THREAD

**Issue by khotilov** [dmlc/xgboost#1200]
I wonder, what were the specific problems with the external memory for R or Windows?

I've added a missing #include to amalgamation, enabled DMLC_ENABLE_STD_THREAD in a few places - and the external memory seems to work everywhere when using gcc 4.8 and above (e.g., gcc 4.9.3 from the most recent Rtools 3.3 in Windows). Granted, I've tried only with fairly small datasets and maybe I didn't yet had a chance to encounter any oddities.

**Comment by tqchen** [comment]
Mainly because the standard Rtools in windows does not come with STD_THREAD support yet, otherwise it is fine.

**Comment by khotilov** [comment]
Well, the std::thread support in the new gcc 4.9.3 Rtools toolchain for R 3.3 was kind of the the main motivation for my attempt to see whether it worked :)

So, should we enable it by default, or do you think some more testing or workarounds might be needed?

**Comment by thirdwing** [comment]
The new toolchain is not stable enough. Please be patient.

**Comment by khotilov** [comment]
@thirdwing : are the following right places for tracking the status of the new Rtools toolchain?
https://github.com/rwinlib/r-base/issues
https://stat.ethz.ch/pipermail/r-sig-windows/
I've been using it for a while since R 3.3 was out, as so far I saw no issues, but I would like to keep an eye.

Regardless of the rtools toolchain, the `#include "../src/data/sparse_page_writer.cc"` include was not in amalgamation, and that was causing the R-package linking issues for me in Linux for some time. I had to either disable `ENABLE_STD_THREAD` in Makevars or to add that line to xgboost-all0.cc. Was that simply a miss which was supposed to be there, or was it not included for some specific reason?
